Crash when executing `Send` within win

Easy to reproduce: start win and just middle click on `Send` in the tag. This is possibly related to #284.

```
panic: acme: no window owner: <nil>


goroutine 47 [running]:
panic(0x87e060, 0xc0005401b0)
	/home/big/go/src/runtime/panic.go:722 +0x2c2 fp=0xc00023b618 sp=0xc00023b588 pc=0x430b72
log.Panicf(0x93606b, 0xd, 0xc00023b700, 0x2, 0x2)
	/home/big/go/src/log/log.go:345 +0xc0 fp=0xc00023b670 sp=0xc00023b618 pc=0x505d40
main.acmeerror(...)
	/home/fhs/go/pkg/mod/github.com/rjkroege/edwood@v0.2.1-0.20190926213824-d3bad36f0fc8/util.go:41
main.(*Window).Eventf(0xc00015a800, 0x9375fa, 0x10, 0xc00023b7b0, 0x5, 0x5)
	/home/fhs/go/pkg/mod/github.com/rjkroege/edwood@v0.2.1-0.20190926213824-d3bad36f0fc8/wind.go:660 +0x371 fp=0xc00023b730 sp=0xc00023b670 pc=0x827d71
main.(*Text).logInsert(0xc00015a918, 0x8, 0xc000014500, 0x2, 0x806)
	/home/fhs/go/pkg/mod/github.com/rjkroege/edwood@v0.2.1-0.20190926213824-d3bad36f0fc8/text.go:462 +0x20b fp=0xc00023b810 sp=0xc00023b730 pc=0x81bb4b
main.(*Text).inserted(0xc00015a918, 0x8, 0xc000014500, 0x2, 0x806)
	/home/fhs/go/pkg/mod/github.com/rjkroege/edwood@v0.2.1-0.20190926213824-d3bad36f0fc8/text.go:441 +0xba fp=0xc00023b850 sp=0xc00023b810 pc=0x81b83a
main.(*File).InsertAt(0xc0000ee500, 0x8, 0xc000014500, 0x2, 0x806)
	/home/fhs/go/pkg/mod/github.com/rjkroege/edwood@v0.2.1-0.20190926213824-d3bad36f0fc8/file.go:338 +0x1d9 fp=0xc00023b8e8 sp=0xc00023b850 pc=0x807799
main.(*Text).Insert(0xc00015a918, 0x8, 0xc000014500, 0x2, 0x806, 0x1)
	/home/fhs/go/pkg/mod/github.com/rjkroege/edwood@v0.2.1-0.20190926213824-d3bad36f0fc8/text.go:481 +0x7f fp=0xc00023b920 sp=0xc00023b8e8 pc=0x81bccf
main.paste(0xc00015a918, 0xc00015a918, 0x0, 0x101, 0x0, 0x0)
	/home/fhs/go/pkg/mod/github.com/rjkroege/edwood@v0.2.1-0.20190926213824-d3bad36f0fc8/exec.go:392 +0x1e1 fp=0xc00023b9c8 sp=0xc00023b920 pc=0x801e91
main.sendx(0xc00015a820, 0xc0003fb820, 0x0, 0x101, 0x0, 0x0)
	/home/fhs/go/pkg/mod/github.com/rjkroege/edwood@v0.2.1-0.20190926213824-d3bad36f0fc8/exec.go:665 +0xa8 fp=0xc00023ba20 sp=0xc00023b9c8 pc=0x803fb8
main.execute(0xc00015a820, 0x4c, 0x4c, 0x1, 0x0)
	/home/fhs/go/pkg/mod/github.com/rjkroege/edwood@v0.2.1-0.20190926213824-d3bad36f0fc8/exec.go:221 +0x419 fp=0xc00023bcb0 sp=0xc00023ba20 pc=0x8002a9
main.xfideventwrite(0xc000528420, 0xc00015a800)
	/home/fhs/go/pkg/mod/github.com/rjkroege/edwood@v0.2.1-0.20190926213824-d3bad36f0fc8/xfid.go:777 +0x3d6 fp=0xc00023bd88 sp=0xc00023bcb0 pc=0x82c156
main.xfidwrite(0xc000528420)
	/home/fhs/go/pkg/mod/github.com/rjkroege/edwood@v0.2.1-0.20190926213824-d3bad36f0fc8/xfid.go:528 +0xb51 fp=0xc00023bf88 sp=0xc00023bd88 pc=0x82aac1
main.xfidctl(0xc000528420, 0x9e62c0, 0xc0000b2208)
	/home/fhs/go/pkg/mod/github.com/rjkroege/edwood@v0.2.1-0.20190926213824-d3bad36f0fc8/xfid.go:39 +0x9c fp=0xc00023bfc8 sp=0xc00023bf88 pc=0x827eac
runtime.goexit()
	/home/big/go/src/runtime/asm_amd64.s:1357 +0x1 fp=0xc00023bfd0 sp=0xc00023bfc8 pc=0x460b81
created by main.xfidallocthread
	/home/fhs/go/pkg/mod/github.com/rjkroege/edwood@v0.2.1-0.20190926213824-d3bad36f0fc8/acme.go:617 +0x221

```

--- a/exec.go
+++ b/exec.go
@@ -21,7 +21,7 @@ import (
 
 type Exectab struct {
 	name  string
-	fn    func(t0, t1, t2 *Text, b0, b1 bool, arg string)
+	fn    func(t, seltext, argt *Text, flag1, flag2 bool, arg string)
 	mark  bool
 	flag1 bool
 	flag2 bool
@@ -120,14 +120,10 @@ func getarg(argt *Text, doaddr bool, dofile bool) (string, string) {
 
 // execute must run with an existing lock on t's Window
 func execute(t *Text, aq0 int, aq1 int, external bool, argt *Text) {
-	var (
-		q0, q1 int
-		r      []rune
-		n, f   int
-	)
+	var n, f int
 
-	q0 = aq0
-	q1 = aq1
+	q0 := aq0
+	q1 := aq1
 	if q1 == q0 { // expand to find word (actually file name)
 		// if in selection, choose selection
 		if t.inSelection(q0) {
@@ -155,7 +151,7 @@ func execute(t *Text, aq0 int, aq1 int, external bool, argt *Text) {
 			}
 		}
 	}
-	r = make([]rune, q1-q0)
+	r := make([]rune, q1-q0)
 	t.file.b.Read(q0, r)
 	e := lookup(string(r))
 	if !external && t.w != nil && t.w.nopen[QWevent] > 0 {
@@ -653,7 +649,8 @@ func run(win *Window, s string, rdir string, newns bool, argaddr string, xarg st
 	go runwaittask(c, cpid)
 }
 
-func sendx(et *Text, _ *Text, _ *Text, _, _ bool, _ string) {
+// sendx appends selected text or snarf buffer to end of body.
+func sendx(et, _, _ *Text, _, _ bool, _ string) {
 	if et.w == nil {
 		return
 	}

--- a/internal/edwoodtest/draw.go
+++ b/internal/edwoodtest/draw.go
@@ -2,7 +2,9 @@
 package edwoodtest
 
 import (
+	"errors"
 	"image"
+	"sync"
 	"unicode/utf8"
 
 	"github.com/rjkroege/edwood/internal/draw"
@@ -11,7 +13,10 @@ import (
 var _ = draw.Display((*mockDisplay)(nil))
 
 // mockDisplay implements draw.Display.
-type mockDisplay struct{}
+type mockDisplay struct {
+	snarfbuf []byte
+	mu       sync.Mutex
+}
 
 // NewDisplay returns a mock draw.Display.
 func NewDisplay() draw.Display {
@@ -31,13 +36,37 @@ func (d *mockDisplay) AllocImage(r image.Rectangle, pix draw.Pix, repl bool, val
 func (d *mockDisplay) AllocImageMix(color1, color3 draw.Color) draw.Image {
 	return NewImage(image.Rectangle{})
 }
-func (d *mockDisplay) Attach(ref int) error                   { return nil }
-func (d *mockDisplay) Flush() error                           { return nil }
-func (d *mockDisplay) ScaleSize(n int) int                    { return 0 }
-func (d *mockDisplay) ReadSnarf(buf []byte) (int, int, error) { return 0, 0, nil }
-func (d *mockDisplay) WriteSnarf(data []byte) error           { return nil }
-func (d *mockDisplay) MoveTo(pt image.Point) error            { return nil }
-func (d *mockDisplay) SetCursor(c *draw.Cursor) error         { return nil }
+func (d *mockDisplay) Attach(ref int) error { return nil }
+func (d *mockDisplay) Flush() error         { return nil }
+func (d *mockDisplay) ScaleSize(n int) int  { return 0 }
+
+// ReadSnarf reads the snarf buffer into buf, returning the number of bytes read,
+// the total size of the snarf buffer (useful if buf is too short), and any
+// error. No error is returned if there is no problem except for buf being too
+// short.
+func (d *mockDisplay) ReadSnarf(buf []byte) (int, int, error) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	n := copy(buf, d.snarfbuf)
+	if n < len(d.snarfbuf) {
+		return n, len(d.snarfbuf), errors.New("short read")
+	}
+	return n, n, nil
+}
+
+// WriteSnarf writes the data to the snarf buffer.
+func (d *mockDisplay) WriteSnarf(data []byte) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	d.snarfbuf = make([]byte, len(data))
+	copy(d.snarfbuf, data)
+	return nil
+}
+
+func (d *mockDisplay) MoveTo(pt image.Point) error    { return nil }
+func (d *mockDisplay) SetCursor(c *draw.Cursor) error { return nil }
 
 var _ = draw.Image((*mockImage)(nil))
 

--- a/xfid.go
+++ b/xfid.go
@@ -706,14 +706,14 @@ func xfideventwrite(x *Xfid, w *Window) {
 	// We can't lock row while we have a window locked
 	// because that can create deadlock with mousethread.
 	rowLock := func() {
-		w.Unlock()
+		defer w.Lock(w.owner)
+		w.Unlock() // sets w.owner to 0
 		row.lk.Lock()
-		w.Lock(w.owner)
 	}
 	rowUnlock := func() {
-		w.Unlock()
+		defer w.Lock(w.owner)
+		w.Unlock() // sets w.owner to 0
 		row.lk.Unlock()
-		w.Lock(w.owner)
 	}
 
 	// The messages have a fixed format: a character indicating the

--- a/xfid_test.go
+++ b/xfid_test.go
@@ -975,6 +975,64 @@ func TestXfidwriteQWevent(t *testing.T) {
 	}
 }
 
+// Issue https://github.com/rjkroege/edwood/issues/285
+func TestXfidwriteQWeventExecuteSend(t *testing.T) {
+	// Setup a new window with "Send" in the tag.
+	d := edwoodtest.NewDisplay()
+	row = Row{
+		display: d,
+	}
+	w := NewWindow().initHeadless(nil)
+	w.col = new(Column)
+	w.nopen[QWevent]++
+	defer func() { w.nopen[QWevent]-- }()
+	w.tag = Text{
+		w: w,
+		file: &File{
+			b:    Buffer("Send"),
+			text: []*Text{&w.tag},
+		},
+		fr:      &MockFrame{},
+		display: d,
+	}
+	w.body = Text{
+		w: w,
+		file: &File{
+			b:    Buffer(""),
+			text: []*Text{&w.body},
+		},
+		fr:      &MockFrame{},
+		display: d,
+	}
+
+	// Put something in the snarf buffer.
+	const snarfbuf = "Hello, 世界\n"
+	d.WriteSnarf([]byte(snarfbuf))
+
+	// Execute "Send" in the tag. This should append the content of
+	// snarf buffer into the body.
+	mr := new(mockResponder)
+	const event = "Mx1 1"
+	x := &Xfid{
+		fcall: plan9.Fcall{
+			Data:  []byte(event),
+			Count: uint32(len(event)),
+		},
+		f: &Fid{
+			qid: plan9.Qid{Path: QID(w.id, QWevent)},
+			w:   w,
+		},
+		fs: mr,
+	}
+	xfidwrite(x)
+	if got := mr.err; got != nil {
+		t.Errorf("event %q: got error %v; want nil", event, got)
+	}
+	if got, want := string(w.body.file.b), snarfbuf; got != want {
+		t.Errorf("body contains %q; want %q", got, want)
+	}
+}
+
 func TestXfidreadEmptyFiles(t *testing.T) {
 	for _, tc := range []struct {
 		name string

